src/autora/state.py

```diff
@@ -45,8 +45,95 @@ def __add__(self: C, other: Union[Delta, Mapping]) -> C:
 S = TypeVar("S", bound=DeltaAddable)
 
 
+class StateDict(UserDict):
```

Comment:
I think this class would benefit from a bunch of doctests which demo and check it's full functionality! You could start with the doctests in the StandardState as an example, and then add the new functionality too.